app/models/deploy_metrics.rb

```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+# Used to collect metrics of cycle time(elapsed time needed to deliver a project to production)
+# Check out https://about.gitlab.com/2019/03/04/reduce-cycle-time/ for more details on cycle time
+# The main objective for this is to gain more insight on the efficiency of the end-to-end project development process
+# There are 2 key metrics:
+# 1. PR - Production: this will show the time needed for a pull request to get deployed to production environment
+# 2. Staging - Production: this will show the time needed to deliver from staging(non-production environment)
+#                          to production environment.
+class DeployMetrics
+  NOT_APPLICABLE = 0
+
+  def initialize(deploy)
+    @deploy = deploy
+  end
+
+  def cycle_time
+    return {} unless @deploy&.production && @deploy&.status == "succeeded"
+
+    {pr_production: pr_production, staging_production: staging_production}
+  end
+
+  private
+
+  # Calculated by averaging deployment end time minus pull requests' creation time
+  def pr_production
+    prs = @deploy.changeset.pull_requests
+    return NOT_APPLICABLE if prs.empty?
+
+    prs.sum { |pr| @deploy.updated_at.to_i - pr.created_at.to_i } / prs.size
+  end
+
+  # Time it took from first staging deploy to finish until the first production deploy finished
+  def staging_production
+    stages = @deploy.project.stages
+    production, staging = stages.partition(&:production?)
+
+    scope = Deploy.successful.where(jobs: {commit: @deploy.commit}).reorder(:id)
+    return NOT_APPLICABLE unless first_staging_deploy = scope.where(stage: staging).first
+    first_production_deploy = scope.where(stage: production).first
```

Comment:
need this or it can be `nil.updated_at` below
```suggestion
    return NOT_APPLICABLE unless first_production_deploy = scope.where(stage: production).first
```

Reply:
if there is no successful production deployment, it seems to me that this line will never be reached ?

Reply:
`first_production_deploy = scope.where(stage: production).first` does not return oO

Reply:
sorry for the confusion, I'm referring to this line
`return {} unless @deploy&.production && @deploy&.status == "succeeded"`

so I assume this `scope` must already have at least 1 production deploy with succeeded status ?

Reply:
ah true!

Reply:
maybe use `.first! # should always be there since we check for that in foo method`